### PR TITLE
feat: add rest endpoint for authorizations' search.

### LINKED ISSUE #21341: Add `Authorization search` REST endpoint
### Description
- [x] Add `/search` endpoint following [the guidelines](https://github.com/camunda/camunda/blob/main/docs/rest-controller.md)

### PR DESCRIPTION
**Pull request by maryarm**
## Description

Add /search endpoint following [the guidelines](https://github.com/camunda/camunda/blob/main/docs/rest-controller.md)

## Related issues

closes #21341

### CODE REVIEW

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 1]
```diff
@@ -1930,6 +1930,64 @@ paths:
             application/problem+json:
               schema:
                 $ref: "#/components/schemas/ProblemDetail"
+  /authoeizations/search:
```
Comment (houssain-barouni) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:1933]:
```suggestion
  /authorizations/search:
```

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 2]
```diff
@@ -1930,6 +1930,64 @@ paths:
...
+          content:
+            application/problem+json:
+              schema:
+                $ref: "#/components/schemas/ProblemDetail"
+        "404":
+          description: "Not found"
+          content:
+            application/problem+json:
+              schema:
+                $ref: "#/components/schemas/ProblemDetail"
```
Comment (houssain-barouni) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:1983]: I am not sure if 403 and 404 errors are returned by this endpoint

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 3]
```diff
@@ -4995,6 +5053,44 @@ components:
...
+    AuthorizationFilterRequest:
+      type: object
+      properties:
+        ownerType:
+          type: "string"
+    AuthorizationResponse:
+      type: "object"
+      properties:
+        ownerKey:
+          type: "string"
```
Comment (houssain-barouni) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:5073]: according to the datamodel, ownerKey should be base64 integer

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 4]
```diff
@@ -4995,6 +5053,44 @@ components:
...
+        ownerType:
+          type: "string"
+        resourceKey:
+          type: "string"
+        resourceType:
+          type: "string"
```
Comment (houssain-barouni) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:5079]: can we a shared enum with `AuthorizationPatchRequest.resourceType` https://github.com/camunda/camunda/blob/2c693cd0ecc0a1c984355761f5cacfcda6c2781c/zeebe/gateway-protocol/src/main/proto/rest-api.yaml#L3500-L3514

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 5]
```diff
@@ -4995,6 +5053,44 @@ components:
...
+        permissions:
+          type: array
+          items:
+            type: string
```
Comment (houssain-barouni) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:5083]: can we have a shared enum with `AuthorizationPatchRequest.permissions.permissionType` https://github.com/camunda/camunda/blob/2c693cd0ecc0a1c984355761f5cacfcda6c2781c/zeebe/gateway-protocol/src/main/proto/rest-api.yaml#L3522-L3526

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 6]
```diff
@@ -4995,6 +5053,44 @@ components:
...
+      properties:
+        ownerType:
+          type: "string"
+    AuthorizationResponse:
+      type: "object"
+      properties:
+        ownerKey:
+          type: "string"
+        ownerType:
+          type: "string"
```
Comment (houssain-barouni) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:5075]: can we have an enum for ownerTypes?

**Review by houssain-barouni — CHANGES_REQUESTED** [review]
Some suggestions on the rest-api specs

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml (hunk 2 above)
Comment (maryarm) [reply zeebe/gateway-protocol/src/main/proto/rest-api.yaml:1983]: done. I've pasted it form users/search before but looking on the code I could find conditions that 403 or 404 could happen.

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml (hunk 3 above)
Comment (maryarm) [reply zeebe/gateway-protocol/src/main/proto/rest-api.yaml]: Could you point me to data model please? as I checked in the code it's string down to AuthorizationEntity.

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 7]
```diff
@@ -3580,6 +3626,83 @@ components:
...
+            format: int64
+        ownerType:
+          enum:
+            - USER
+            - ROLE
+            - GROUP
+            - MAPPING
+            - UNSPECIFIED
+        resourceType:
+          description: The type of resource to add/remove perissions to/from.
```
Comment (christinaausley) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:3664]: Typo -- permissions 👍

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 8]
```diff
@@ -5044,6 +5102,44 @@ components:
...
+        - $ref: "#/components/schemas/SearchQueryRequest"
+      type: object
+      properties:
+        filter:
+          allOf:
+            - $ref: "#/components/schemas/AuthorizationFilterRequest"
+    AuthorizationFilterRequest:
+      type: object
+      properties:
+        ownerType:
```
Comment (tmetzke) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml]: 🔧 Let's add descriptions to properties in requests and responses so users can understand what they can use this for and what those properties mean.

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 9]
```diff
@@ -1970,6 +1970,64 @@ paths:
...
+      summary: "Query authorizations (alpha)"
+      description: |
+        Search for authorizations based on given criteria.
+
+        :::note
+        This endpoint is an alpha feature and not enabled on Camunda clusters out of the box.
+        See the [Camunda 8 REST API overview](/apis-tools/camunda-api-rest/camunda-api-rest-overview.md#query-api)
+        for further details.
+        :::
+      operationId: "findAllAuthorizations"
```
Comment (tmetzke) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:1986]: 🔧 For consistency with the other search endpoints, can we name this `findAuthorizations`? 🙂

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 10]
```diff
@@ -3580,6 +3626,83 @@ components:
...
+        ownerKey:
+            type: integer
+            format: int64
+        ownerType:
+          enum:
+            - USER
+            - ROLE
+            - GROUP
+            - MAPPING
+            - UNSPECIFIED
```
Comment (tmetzke) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:3662]: 🔧 Let's extract the enum so we don't repeat ourselves in the request and response. I am also rather sure that we will need this enum in other entities soon 🙂

File: zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/AuthorizationQueryControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+        .isBadRequest()
+        .expectHeader()
+        .contentType(MediaType.APPLICATION_PROBLEM_JSON)
+        .expectBody()
+        .json(expectedResponse);
+
+    verify(authorizationServices, never()).search(any(AuthorizationQuery.class));
+  }
+
+  public static Stream<Arguments> invalidAuthorizationSearchQueries() {
```
Comment (tmetzke) [inline zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/AuthorizationQueryControllerTest.java:196]: 👍 This is really nice. You get a direct overview of bad queries we test and what the result will be, love it 🦖

**Review by tmetzke — APPROVED** [review]
👍 Really good work, @maryarm 🚀 
👍 The code looks really good, the controller, mappers, and tests look flawless ⭐ 
🔧 There are some slight improvements we can make to the OpenAPI spec but it's really minimal.

Pre-approving, assuming you will tend to the minor suggestions that were raised. Really well done!

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 12]
```diff
@@ -3580,6 +3626,83 @@ components:
                 description: A list of resource IDs the permission relates to.
                 items:
                   type: string
+    AuthorizationSearchQueryRequest:
+      allOf:
+        - $ref: "#/components/schemas/SearchQueryRequest"
+      type: object
+      properties:
+        filter:
```
Comment (pepopowitz) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:3634]: 🔧 Please add a description to this property.

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 13]
```diff
@@ -3580,6 +3626,83 @@ components:
...
+          allOf:
+            - $ref: "#/components/schemas/AuthorizationFilterRequest"
+    AuthorizationFilterRequest:
+      type: object
+      properties:
```
Comment (pepopowitz) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:3639]: 🔧 Please add descriptions to the properties on this schema.

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 14]
```diff
@@ -3580,6 +3626,83 @@ components:
...
+            - ROLE
+            - GROUP
+            - MAPPING
+            - UNSPECIFIED
+        ownerKey:
+          type: integer
+          format: int64
+    AuthorizationResponse:
+      type: "object"
+      properties:
```
Comment (pepopowitz) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:3652]: 🔧 Please add descriptions to all properties on this schema.

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml [hunk 15]
```diff
@@ -3580,6 +3626,83 @@ components:
...
+              resourceIds:
+                type: array
+                description: A list of resource IDs the permission relates to.
+                items:
+                  type: string
+    AuthorizationSearchResponse:
+      type: object
+      allOf:
+        - $ref: "#/components/schemas/SearchQueryResponse"
+      properties:
```
Comment (pepopowitz) [inline zeebe/gateway-protocol/src/main/proto/rest-api.yaml:3701]: 🔧 Please add a description to the property on this schema.

**Review by pepopowitz — APPROVED** [review]
🔧 Some missing descriptions, but otherwise the spec looks good from my end!

File: zeebe/gateway-protocol/src/main/proto/rest-api.yaml (hunk 3 above)
Comment (houssain-barouni) [reply zeebe/gateway-protocol/src/main/proto/rest-api.yaml]: I was looking at https://github.com/camunda/camunda/blob/main/webapps-schema/src/main/java/io/camunda/webapps/schema/entities/usermanagement/AuthorizationEntity.java, there is indeed inconsistency between the entity used by the exporter and the one used in the search

**Review by houssain-barouni — APPROVED** [review]
LGTM

**Pull request merged**